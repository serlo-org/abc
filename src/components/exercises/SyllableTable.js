import React from 'react';
import { addIndex, map } from 'ramda';
import { View } from 'react-native';

import { PRIMARY } from '../../styles/colors';
import { loadSound } from '../helpers/audio';
import { RoundTextButton, RoundText } from '../Components';

const mapIndexed = addIndex(map);

const SyllableRow = ({
  letter,
  letterKey,
  sound,
  vowels,
  letterIndex,
  vowelIndex
}) => {
  const play = () => {
    sound.playAsync();
    sound.setPlaybackFinishedCallback(() => {
      sound.setPositionAsync(0);
    });
  };

  return (
    <View
      style={{
        alignItems: 'center',
        justifyContent: 'center',
        flexDirection: 'row'
      }}
    >
      {mapIndexed((vowel, vowelKey) => {
        const syllable = letter + vowel;
        if (letterKey === letterIndex && vowelKey === vowelIndex) {
          return (
            <RoundTextButton
              key={vowelKey}
              style={{ margin: 5 }}
              text={syllable}
              size={60}
              onPress={play}
            />
          );
        }

        return (
          <RoundText
            key={vowelKey}
            style={{ margin: 5, opacity: 0.4 }}
            text={syllable}
            size={60}
          />
        );
      }, vowels)}
    </View>
  );
};

const SyllableTable = ({ letters, ...props }) => {
  return (
    <View
      style={{
        flex: 1,
        backgroundColor: PRIMARY,
        alignItems: 'center'
      }}
    >
      <View
        style={{
          alignItems: 'center',
          justifyContent: 'center',
          flexDirection: 'column',
          marginTop: 50
        }}
      >
        {mapIndexed(
          (letter, letterKey) => (
            <SyllableRow
              key={letterKey}
              letter={letter}
              letterKey={letterKey}
              {...props}
            />
          ),
          letters
        )}
      </View>
    </View>
  );
};

export default loadSound(SyllableTable);
