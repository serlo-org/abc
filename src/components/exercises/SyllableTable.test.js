import React from 'react';

import SyllableTable from './SyllableTable';

import renderer from 'react-test-renderer';

it('renders without crashing (3x3)', () => {
  const tree = renderer.create(
    <SyllableTable
      sound={require('../../assets/sounds/ne.mp3')}
      letters={['n', 's', 't']}
      vowels={['a', 'e', 'o']}
      letterIndex={0}
      vowelIndex={1}
    />
  );
  expect(tree).toMatchSnapshot();
});

it('renders without crashing (5x4)', () => {
  const tree = renderer.create(
    <SyllableTable
      sound={require('../../assets/sounds/ha.mp3')}
      letters={['h', 'n', 'r', 's', 't']}
      vowels={['a', 'e', 'i', 'o']}
      letterIndex={0}
      vowelIndex={0}
    />
  );
  expect(tree).toMatchSnapshot();
});
