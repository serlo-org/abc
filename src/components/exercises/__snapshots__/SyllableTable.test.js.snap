// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#00B4D5",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "center",
        "marginTop": 50,
        "width": 300,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255,255,255,0.2)",
            "borderRadius": 9999,
            "height": 60,
            "justifyContent": "center",
            "padding": 5,
            "width": 60,
          },
          Object {
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#fff",
              "fontSize": 40,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        na
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "margin": 5,
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(255,255,255,0.2)",
              "borderRadius": 9999,
              "height": 60,
              "justifyContent": "center",
              "padding": 5,
              "width": 60,
            },
            Object {
              "backgroundColor": "#73DBFF",
              "elevation": 5,
              "opacity": 1,
              "shadowColor": "rgba(0, 0, 0, 0.1)",
              "shadowOffset": Object {
                "height": 4,
                "width": 4,
              },
              "shadowOpacity": 1,
              "shadowRadius": 0,
            },
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "color": "#fff",
                "fontSize": 40,
                "fontWeight": "bold",
                "textAlign": "center",
              },
              undefined,
            ]
          }
        >
          ne
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255,255,255,0.2)",
            "borderRadius": 9999,
            "height": 60,
            "justifyContent": "center",
            "padding": 5,
            "width": 60,
          },
          Object {
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#fff",
              "fontSize": 40,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        no
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255,255,255,0.2)",
            "borderRadius": 9999,
            "height": 60,
            "justifyContent": "center",
            "padding": 5,
            "width": 60,
          },
          Object {
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#fff",
              "fontSize": 40,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        sa
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255,255,255,0.2)",
            "borderRadius": 9999,
            "height": 60,
            "justifyContent": "center",
            "padding": 5,
            "width": 60,
          },
          Object {
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#fff",
              "fontSize": 40,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        se
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255,255,255,0.2)",
            "borderRadius": 9999,
            "height": 60,
            "justifyContent": "center",
            "padding": 5,
            "width": 60,
          },
          Object {
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#fff",
              "fontSize": 40,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        so
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255,255,255,0.2)",
            "borderRadius": 9999,
            "height": 60,
            "justifyContent": "center",
            "padding": 5,
            "width": 60,
          },
          Object {
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#fff",
              "fontSize": 40,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        ta
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255,255,255,0.2)",
            "borderRadius": 9999,
            "height": 60,
            "justifyContent": "center",
            "padding": 5,
            "width": 60,
          },
          Object {
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#fff",
              "fontSize": 40,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        te
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255,255,255,0.2)",
            "borderRadius": 9999,
            "height": 60,
            "justifyContent": "center",
            "padding": 5,
            "width": 60,
          },
          Object {
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#fff",
              "fontSize": 40,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        to
      </Text>
    </View>
  </View>
</View>
`;
